---

title: Design


keywords: fastai
sidebar: home_sidebar

summary: "Design sequences for regions of interest"
description: "Design sequences for regions of interest"
nb_path: "02_design.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 02_design.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="reverse_compliment" class="doc_header"><code>reverse_compliment</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L14" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>reverse_compliment</code>(<strong><code>seq</code></strong>)</p>
</blockquote>
<p>Return the reverse compliment of a sequence</p>
<p>seq: str |</p>
<p>return: str</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="n">reverse_compliment</span><span class="p">(</span><span class="s1">&#39;ACTG&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;CAGT&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Global-indexing">Global indexing<a class="anchor-link" href="#Global-indexing"> </a></h2><p>Given the global start and end of a sequence (location with the genome) and its orientation, track the
the global position of targetting sgRNAs.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="calculate_global_position" class="doc_header"><code>calculate_global_position</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L27" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>calculate_global_position</code>(<strong><code>strand</code></strong>, <strong><code>start</code></strong>, <strong><code>end</code></strong>, <strong><code>relative_position</code></strong>)</p>
</blockquote>
<p>Calculate the global position of locus in a subsequence</p>
<p>strand: int [1 or -1], strand of DNA sequence |
start: int, global start position of sequence |
end: int, global end position of sequence |
relative_position: list of int, positions of sgRNAs relative to sequence |</p>
<p>returns: list of int</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">region_start</span> <span class="o">=</span> <span class="mi">20100</span>
<span class="n">region_end</span> <span class="o">=</span> <span class="mi">20200</span>
<span class="n">chrom</span> <span class="o">=</span> <span class="s1">&#39;1&#39;</span>
<span class="n">sequence</span> <span class="o">=</span> <span class="n">ensembl</span><span class="o">.</span><span class="n">get_region_sequence</span><span class="p">(</span><span class="n">region_start</span><span class="p">,</span> <span class="n">region_end</span><span class="p">,</span> <span class="n">chrom</span><span class="p">)</span>
<span class="n">sgrna_df</span> <span class="o">=</span> <span class="n">tile</span><span class="o">.</span><span class="n">build_sgrna_df</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="o">-</span><span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="n">pams</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CGG&#39;</span><span class="p">,</span> <span class="s1">&#39;AGG&#39;</span><span class="p">,</span> <span class="s1">&#39;TGG&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">])</span>
<span class="n">sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">calculate_global_position</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">region_start</span><span class="p">,</span> <span class="n">region_end</span><span class="p">,</span> <span class="n">sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_relative_start&#39;</span><span class="p">])</span>
<span class="k">assert</span><span class="p">(</span><span class="n">sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_sequence&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">ensembl</span><span class="o">.</span><span class="n">get_region_sequence</span><span class="p">(</span><span class="n">sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_start&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                                                                    <span class="n">sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_start&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">19</span><span class="p">,</span>
                                                                    <span class="n">chrom</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Note that the direction is reversed for a sequence on the negative DNA strand</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">reverse_sequence</span> <span class="o">=</span> <span class="n">reverse_compliment</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
<span class="n">rev_sgrna_df</span> <span class="o">=</span> <span class="n">tile</span><span class="o">.</span><span class="n">build_sgrna_df</span><span class="p">(</span><span class="n">reverse_sequence</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="o">-</span><span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="n">pams</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CGG&#39;</span><span class="p">,</span> <span class="s1">&#39;AGG&#39;</span><span class="p">,</span> <span class="s1">&#39;TGG&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">])</span>
<span class="n">rev_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">calculate_global_position</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">region_start</span><span class="p">,</span> <span class="n">region_end</span><span class="p">,</span>
                                                               <span class="n">rev_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_relative_start&#39;</span><span class="p">])</span>
<span class="k">assert</span><span class="p">(</span><span class="n">rev_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_sequence&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">reverse_compliment</span><span class="p">(</span><span class="n">ensembl</span><span class="o">.</span><span class="n">get_region_sequence</span><span class="p">(</span>
    <span class="n">rev_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_start&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">19</span><span class="p">,</span>
    <span class="n">rev_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_start&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">chrom</span><span class="p">)))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="traverse_global_position" class="doc_header"><code>traverse_global_position</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L46" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>traverse_global_position</code>(<strong><code>strand</code></strong>, <strong><code>reference_position</code></strong>, <strong><code>distance</code></strong>)</p>
</blockquote>
<p>Move from one global position to another</p>
<p>strand: int [1 or -1], strand of DNA sequence |
reference_position: list of int, global starting position |
distance: int, distance to move 5' to 3' on either strand |</p>
<p>returns: list of int</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">sgrna_positions</span> <span class="o">=</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">]</span>
<span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">sgrna_positions</span><span class="p">:</span>
    <span class="n">sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pos</span><span class="p">)]</span> <span class="o">=</span> <span class="n">traverse_global_position</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_start&#39;</span><span class="p">],</span> <span class="n">pos</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="k">assert</span> <span class="p">(</span><span class="n">sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_sequence&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span> <span class="o">==</span> <span class="n">ensembl</span><span class="o">.</span><span class="n">get_region_sequence</span><span class="p">(</span><span class="n">sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_4&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                                                                          <span class="n">sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_8&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                                                                          <span class="n">chrom</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Again the direction is reversed for a sequence on the negative DNA strand</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">sgrna_positions</span><span class="p">:</span>
    <span class="n">rev_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pos</span><span class="p">)]</span> <span class="o">=</span> <span class="n">traverse_global_position</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">rev_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_start&#39;</span><span class="p">],</span> <span class="n">pos</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">rev_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_sequence&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span> <span class="o">==</span> <span class="n">reverse_compliment</span><span class="p">(</span><span class="n">ensembl</span><span class="o">.</span><span class="n">get_region_sequence</span><span class="p">(</span>
    <span class="n">rev_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_8&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">rev_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_global_4&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">chrom</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_sgrna_global_indices" class="doc_header"><code>get_sgrna_global_indices</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L64" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_sgrna_global_indices</code>(<strong><code>sgrna_df</code></strong>, <strong><code>seq_start</code></strong>, <strong><code>seq_end</code></strong>, <strong><code>strand</code></strong>, <strong><code>sg_positions</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Take an sgrna_df and return the globabl positions of elements of the sgRNA</p>
<p>sgrna_df: Dataframe from <a href="/sgrna_designer/tile.html#build_sgrna_df"><code>build_sgrna_df</code></a> |
seq_start: int, starting index of sequence |
seq_end: int, ending index of sequence |
strand: int [1 or -1], strand of DNA sequence |
sg_positions: list, positions within the sgRNA to annotate (e.g. [4,8] for nucleotides 4 and 8 of the sgRNA) |</p>
<p>returns: DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Building-a-full-sgRNA-design-run">Building a full sgRNA design run<a class="anchor-link" href="#Building-a-full-sgRNA-design-run"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_trainscript_region_info" class="doc_header"><code>get_trainscript_region_info</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L87" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_trainscript_region_info</code>(<strong><code>transcript_info</code></strong>, <strong><code>region_parent</code></strong>, <strong><code>region</code></strong>)</p>
</blockquote>
<p>Return dictionaries for specified regions of interest</p>
<p>transcript_info: dict, returned by <a href="https://rest.ensembl.org/documentation/info/lookup">https://rest.ensembl.org/documentation/info/lookup</a> |
region_parent: str, first level key in transcript_info eg. UTR |
region: str, second level key in trancript_info eg. three_prime_UTR |</p>
<p>returns: list of dict</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_target_regions_df" class="doc_header"><code>get_target_regions_df</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L106" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_target_regions_df</code>(<strong><code>target_transcripts</code></strong>, <strong><code>region_parent</code></strong>, <strong><code>region</code></strong>, <strong><code>expand_3prime</code></strong>, <strong><code>expand_5prime</code></strong>)</p>
</blockquote>
<p>For a list of transcripts return the locations for a specific region (e.g. three_prime_UTR)</p>
<p>target_transcripts: list of str
region_parent: str, first level key in transcript_info eg. UTR |
region: str, second level key in trancript_info eg. three_prime_UTR |
expand_3prime: int, length to expand region in 3' direction |
expand_5prime: int, length to expand region in 5' direction |</p>
<p>returns: DataFrame of target regions</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">regions</span> <span class="o">=</span> <span class="n">get_target_regions_df</span><span class="p">([</span><span class="s1">&#39;ENST00000381577&#39;</span><span class="p">],</span> <span class="s1">&#39;UTR&#39;</span><span class="p">,</span> <span class="s1">&#39;five_prime_UTR&#39;</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">30</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">regions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span>
<span class="k">assert</span> <span class="n">regions</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">5450542</span>
<span class="k">assert</span> <span class="n">regions</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">5450596</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_target_regions_sequences" class="doc_header"><code>get_target_regions_sequences</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L137" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_target_regions_sequences</code>(<strong><code>target_regions_df</code></strong>)</p>
</blockquote>
<p>Get sequences from a DataFrame of target regions</p>
<p>target_regions_df: DataFrame from <a href="/sgrna_designer/design.html#get_target_regions_df"><code>get_target_regions_df</code></a></p>
<p>returns: DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">target_sequences_df</span> <span class="o">=</span> <span class="n">get_target_regions_sequences</span><span class="p">(</span><span class="n">regions</span><span class="p">)</span>
<span class="k">assert</span> <span class="s1">&#39;AGTTCTGCGCAGCTTCCCGAGGCTCCGCACCAGCCGCGCTTCTGTCCGCCTGCAG&#39;</span> <span class="ow">in</span> <span class="n">target_sequences_df</span><span class="p">[</span><span class="s1">&#39;seq&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="merge_target_region_sequence_df" class="doc_header"><code>merge_target_region_sequence_df</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L151" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>merge_target_region_sequence_df</code>(<strong><code>target_regions_df</code></strong>, <strong><code>target_sequences_df</code></strong>)</p>
</blockquote>
<p>Merge target regions and target sequences dataframes</p>
<p>target_regions_df:  DataFrame, from <a href="/sgrna_designer/design.html#get_target_regions_df"><code>get_target_regions_df</code></a> |
target_sequences_df: DataFrame, from <a href="/sgrna_designer/design.html#get_target_regions_sequences"><code>get_target_regions_sequences</code></a> |</p>
<p>returns: DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">target_region_seq_df</span> <span class="o">=</span> <span class="n">merge_target_region_sequence_df</span><span class="p">(</span><span class="n">regions</span><span class="p">,</span> <span class="n">target_sequences_df</span><span class="p">)</span>
<span class="n">target_region_seq_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>db_type</th>
      <th>Parent</th>
      <th>region_id</th>
      <th>end</th>
      <th>assembly_name</th>
      <th>object_type</th>
      <th>start</th>
      <th>species</th>
      <th>source</th>
      <th>seq_region_name</th>
      <th>strand</th>
      <th>transcript_id</th>
      <th>expanded_start</th>
      <th>expanded_end</th>
      <th>region_pos</th>
      <th>region_pos_id</th>
      <th>molecule</th>
      <th>seq</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>core</td>
      <td>ENST00000381577</td>
      <td>ENST00000381577</td>
      <td>5450596</td>
      <td>GRCh38</td>
      <td>five_prime_UTR</td>
      <td>5450542</td>
      <td>homo_sapiens</td>
      <td>ensembl_havana</td>
      <td>9</td>
      <td>1</td>
      <td>ENST00000381577</td>
      <td>5450512</td>
      <td>5450626</td>
      <td>9:5450512..5450626</td>
      <td>chromosome:GRCh38:9:5450512:5450626:1</td>
      <td>dna</td>
      <td>CTGAGCAGCTGGCGCGTCCCGCGCGGCCCCAGTTCTGCGCAGCTTC...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>core</td>
      <td>ENST00000381577</td>
      <td>ENST00000381577</td>
      <td>5456113</td>
      <td>GRCh38</td>
      <td>five_prime_UTR</td>
      <td>5456100</td>
      <td>homo_sapiens</td>
      <td>ensembl_havana</td>
      <td>9</td>
      <td>1</td>
      <td>ENST00000381577</td>
      <td>5456070</td>
      <td>5456143</td>
      <td>9:5456070..5456143</td>
      <td>chromosome:GRCh38:9:5456070:5456143:1</td>
      <td>dna</td>
      <td>CAGTCTTCTTTTCGTGTTTTCCATAATTAGGGCATTCCAGAAAGAT...</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_transcript_sgrnas" class="doc_header"><code>get_transcript_sgrnas</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L167" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_transcript_sgrnas</code>(<strong><code>target_region_seq_df</code></strong>, <strong><code>context_len</code></strong>, <strong><code>pam_start</code></strong>, <strong><code>pam_len</code></strong>, <strong><code>sgrna_start</code></strong>, <strong><code>sgrna_len</code></strong>, <strong><code>pams</code></strong>, <strong><code>sg_positions</code></strong>)</p>
</blockquote>
<p>Get all sgRNAs targeting transcript regions</p>
<p>target_region_seq_df: DataFrame, from <a href="/sgrna_designer/design.html#merge_target_region_sequence_df"><code>merge_target_region_sequence_df</code></a>
context_len: int, length of context sequence |
pam_start: int, position of PAM start relative to the context sequence |
pam_len: int, length of PAM |
sgrna_start: int, position of sgRNA relative to context sequence |
sgrna_len: length of sgRNA sequence |
pams: list or None, PAMs to design against |
sg_positions: list of int, positions within the sgRNA to annotate
(e.g. [4,8] for nucleotides 4 and 8 of the sgRNA) |</p>
<p>returns DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">transcript_sgrna_df</span> <span class="o">=</span> <span class="n">get_transcript_sgrnas</span><span class="p">(</span><span class="n">target_region_seq_df</span><span class="p">,</span> <span class="n">context_len</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">pam_start</span><span class="o">=-</span><span class="mi">6</span><span class="p">,</span>
                                            <span class="n">pam_len</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">sgrna_start</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">sgrna_len</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">pams</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;AGG&#39;</span><span class="p">,</span> <span class="s1">&#39;CGG&#39;</span><span class="p">,</span> <span class="s1">&#39;TGG&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">],</span>
                                            <span class="n">sg_positions</span><span class="o">=</span><span class="p">[</span><span class="mi">17</span><span class="p">,</span> <span class="mi">18</span><span class="p">])</span>
<span class="k">assert</span> <span class="s1">&#39;CCGCGCTTCTGTCCGCCTGC&#39;</span> <span class="ow">in</span> <span class="n">transcript_sgrna_df</span><span class="p">[</span><span class="s1">&#39;sgrna_sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="filter_sgrnas_by_region" class="doc_header"><code>filter_sgrnas_by_region</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L216" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>filter_sgrnas_by_region</code>(<strong><code>transcript_sgrna_df</code></strong>, <strong><code>sg_positions</code></strong>)</p>
</blockquote>
<p>Filter for sgRNAs such that the specified sgRNA positions are within the region of interest</p>
<p>transcript_sgrna_df: DataFrame, from<a href="/sgrna_designer/design.html#get_transcript_sgrnas"><code>get_transcript_sgrnas</code></a>
sg_positions: list of int, positions within the sgRNA to annotate
(e.g. [4,8] for nucleotides 4 and 8 of the sgRNA) |</p>
<p>returns: DataFrame, filtered for sgRNAs that fall within the target region</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">filtered_sgrnas</span> <span class="o">=</span> <span class="n">filter_sgrnas_by_region</span><span class="p">(</span><span class="n">transcript_sgrna_df</span><span class="p">,</span> <span class="p">[</span><span class="mi">17</span><span class="p">,</span> <span class="mi">18</span><span class="p">])</span>
<span class="k">assert</span> <span class="n">filtered_sgrnas</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">transcript_sgrna_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="flag_sgrnas" class="doc_header"><code>flag_sgrnas</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L237" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>flag_sgrnas</code>(<strong><code>sgrna_df</code></strong>, <strong><code>flag_seqs</code></strong>, <strong><code>flag_seqs_start</code></strong>, <strong><code>flag_seqs_end</code></strong>)</p>
</blockquote>
<p>Flag problematic subsequences (e.g. TTTT) in sgRNAs</p>
<p>sgrna_df: DataFrame |
flag_seqs: list of str, flag subsequences anywhere in sgRNA sequences |
flag_seqs_start: list of str, flag subsequences at the beginning of sgRNA sequences |
flag_seqs_end: list of str, flag subsequences at the end of sgRNA sequences |</p>
<p>returns: DataFrame, with additional columns to flag problematic sgRNAs</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">problematic_sgrna_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;sgrna_sequence&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;TCTCTTGGAATTGGTGGTGG&#39;</span><span class="p">,</span> <span class="s1">&#39;CCTTACTGAGATCTGGAGAC&#39;</span><span class="p">,</span>
                                                        <span class="s1">&#39;GGATTTTGAAGGAGACGGAC&#39;</span><span class="p">,</span> <span class="s1">&#39;AGACGGACTGGTGAGAATTT&#39;</span><span class="p">,</span>
                                                        <span class="s1">&#39;TCTCGCCTCTATTGAGCTGC&#39;</span><span class="p">,</span> <span class="s1">&#39;TCTCCAGATAATTTTTATGG&#39;</span><span class="p">,</span>
                                                        <span class="s1">&#39;GAGACGGACTGGTGAGAATT&#39;</span><span class="p">,</span> <span class="s1">&#39;TCTCATCTCTTTTCTTTTTG&#39;</span><span class="p">,</span>
                                                        <span class="s1">&#39;AGAATTGGATCTGGATCATT&#39;</span><span class="p">,</span> <span class="s1">&#39;GGTTCCCAAGGACCTATATG&#39;</span><span class="p">,</span>
                                                        <span class="s1">&#39;AGATTTTCTACTGCACTTTT&#39;</span><span class="p">,</span> <span class="s1">&#39;ATAATTTTTATGGTGGGACG&#39;</span><span class="p">]})</span>

<span class="n">flagged_sgrnas</span> <span class="o">=</span> <span class="n">flag_sgrnas</span><span class="p">(</span><span class="n">problematic_sgrna_df</span><span class="p">,</span> <span class="n">flag_seqs</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TTTT&#39;</span><span class="p">,</span> <span class="s1">&#39;CGTCTC&#39;</span><span class="p">,</span> <span class="s1">&#39;GAGACG&#39;</span><span class="p">],</span>
                             <span class="n">flag_seqs_start</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TCTC&#39;</span><span class="p">,</span> <span class="s1">&#39;AGACG&#39;</span><span class="p">],</span> <span class="n">flag_seqs_end</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GAGAC&#39;</span><span class="p">])</span>
<span class="k">assert</span> <span class="p">(</span><span class="n">flagged_sgrnas</span><span class="p">[</span><span class="s1">&#39;flag&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">==</span> <span class="mi">10</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="design_sgrna_tiling_library" class="doc_header"><code>design_sgrna_tiling_library</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/design.py#L266" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>design_sgrna_tiling_library</code>(<strong><code>target_transcripts</code></strong>, <strong><code>region_parent</code></strong>, <strong><code>region</code></strong>, <strong><code>expand_3prime</code></strong>, <strong><code>expand_5prime</code></strong>, <strong><code>context_len</code></strong>, <strong><code>pam_start</code></strong>, <strong><code>pam_len</code></strong>, <strong><code>sgrna_start</code></strong>, <strong><code>sgrna_len</code></strong>, <strong><code>pams</code></strong>, <strong><code>sg_positions</code></strong>, <strong><code>flag_seqs</code></strong>, <strong><code>flag_seqs_start</code></strong>, <strong><code>flag_seqs_end</code></strong>)</p>
</blockquote>
<p>Design sgRNAs tiling transcript regions</p>
<p>target_transcripts: list of str
region_parent: str, first level key in transcript_info eg. UTR |
region: str, second level key in trancript_info eg. three_prime_UTR |
expand_3prime: int, length to expand region in 3' direction |
expand_5prime: int, length to expand region in 5' direction |
context_len: int, length of context sequence |
pam_start: int, position of PAM start relative to the context sequence |
pam_len: int, length of PAM |
sgrna_start: int, position of sgRNA relative to context sequence |
sgrna_len: length of sgRNA sequence |
pams: list or None, PAMs to design against |
sg_positions: list of int, positions within the sgRNA to annotate
    (e.g. [4,8] for nucleotides 4 and 8 of the sgRNA) |
flag_seqs: list of str, flag subsequences anywhere in sgRNA sequences |
flag_seqs_start: list of str, flag subsequences at the beginning of sgRNA sequences |
flag_seqs_end: list of str, flag subsequences at the end of sgRNA sequences |</p>
<p>returns: DataFrame, sgRNAs for transcript regions of interest</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">target_transcripts</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ENST00000381577&#39;</span><span class="p">,</span> <span class="s1">&#39;ENST00000644969&#39;</span><span class="p">]</span>
<span class="n">exon_tiling_library</span> <span class="o">=</span> <span class="n">design_sgrna_tiling_library</span><span class="p">(</span><span class="n">target_transcripts</span><span class="p">,</span> <span class="n">region_parent</span><span class="o">=</span><span class="s1">&#39;Exon&#39;</span><span class="p">,</span> <span class="n">region</span><span class="o">=</span><span class="s1">&#39;Exon&#39;</span><span class="p">,</span>
                                                  <span class="n">expand_3prime</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">expand_5prime</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">context_len</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
                                                  <span class="n">pam_start</span><span class="o">=-</span><span class="mi">6</span><span class="p">,</span> <span class="n">pam_len</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">sgrna_start</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">sgrna_len</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
                                                  <span class="n">pams</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;AGG&#39;</span><span class="p">,</span> <span class="s1">&#39;CGG&#39;</span><span class="p">,</span> <span class="s1">&#39;TGG&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">],</span> <span class="n">sg_positions</span><span class="o">=</span><span class="p">[</span><span class="mi">17</span><span class="p">,</span> <span class="mi">18</span><span class="p">],</span>
                                                  <span class="n">flag_seqs</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TTTT&#39;</span><span class="p">,</span> <span class="s1">&#39;CGTCTC&#39;</span><span class="p">,</span> <span class="s1">&#39;GAGACG&#39;</span><span class="p">],</span>
                                                  <span class="n">flag_seqs_start</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TCTC&#39;</span><span class="p">,</span> <span class="s1">&#39;AGACG&#39;</span><span class="p">],</span> <span class="n">flag_seqs_end</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GAGAC&#39;</span><span class="p">])</span>
<span class="n">utr3_tiling_library</span> <span class="o">=</span> <span class="n">design_sgrna_tiling_library</span><span class="p">(</span><span class="n">target_transcripts</span><span class="p">,</span> <span class="n">region_parent</span><span class="o">=</span><span class="s1">&#39;UTR&#39;</span><span class="p">,</span> <span class="n">region</span><span class="o">=</span><span class="s1">&#39;three_prime_UTR&#39;</span><span class="p">,</span>
                                                  <span class="n">expand_3prime</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">expand_5prime</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">context_len</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
                                                  <span class="n">pam_start</span><span class="o">=-</span><span class="mi">6</span><span class="p">,</span> <span class="n">pam_len</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">sgrna_start</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">sgrna_len</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
                                                  <span class="n">pams</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;AGG&#39;</span><span class="p">,</span> <span class="s1">&#39;CGG&#39;</span><span class="p">,</span> <span class="s1">&#39;TGG&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">],</span> <span class="n">sg_positions</span><span class="o">=</span><span class="p">[</span><span class="mi">17</span><span class="p">,</span> <span class="mi">18</span><span class="p">],</span>
                                                  <span class="n">flag_seqs</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TTTT&#39;</span><span class="p">,</span> <span class="s1">&#39;CGTCTC&#39;</span><span class="p">,</span> <span class="s1">&#39;GAGACG&#39;</span><span class="p">],</span>
                                                  <span class="n">flag_seqs_start</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TCTC&#39;</span><span class="p">,</span> <span class="s1">&#39;AGACG&#39;</span><span class="p">],</span> <span class="n">flag_seqs_end</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GAGAC&#39;</span><span class="p">])</span>
<span class="n">utr5_tiling_library</span> <span class="o">=</span> <span class="n">design_sgrna_tiling_library</span><span class="p">(</span><span class="n">target_transcripts</span><span class="p">,</span> <span class="n">region_parent</span><span class="o">=</span><span class="s1">&#39;UTR&#39;</span><span class="p">,</span> <span class="n">region</span><span class="o">=</span><span class="s1">&#39;five_prime_UTR&#39;</span><span class="p">,</span>
                                                  <span class="n">expand_3prime</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">expand_5prime</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">context_len</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
                                                  <span class="n">pam_start</span><span class="o">=-</span><span class="mi">6</span><span class="p">,</span> <span class="n">pam_len</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">sgrna_start</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">sgrna_len</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
                                                  <span class="n">pams</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;AGG&#39;</span><span class="p">,</span> <span class="s1">&#39;CGG&#39;</span><span class="p">,</span> <span class="s1">&#39;TGG&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">],</span> <span class="n">sg_positions</span><span class="o">=</span><span class="p">[</span><span class="mi">17</span><span class="p">,</span> <span class="mi">18</span><span class="p">],</span>
                                                  <span class="n">flag_seqs</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TTTT&#39;</span><span class="p">,</span> <span class="s1">&#39;CGTCTC&#39;</span><span class="p">,</span> <span class="s1">&#39;GAGACG&#39;</span><span class="p">],</span>
                                                  <span class="n">flag_seqs_start</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TCTC&#39;</span><span class="p">,</span> <span class="s1">&#39;AGACG&#39;</span><span class="p">],</span> <span class="n">flag_seqs_end</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GAGAC&#39;</span><span class="p">])</span>
<span class="c1"># This is an approximation, as some guides are actually at the UTR/CDS border</span>
<span class="n">cds_tiling_library</span> <span class="o">=</span> <span class="n">exon_tiling_library</span><span class="p">[</span><span class="o">~</span><span class="n">exon_tiling_library</span><span class="p">[</span><span class="s1">&#39;sgrna_sequence&#39;</span><span class="p">]</span>
                                         <span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">utr3_tiling_library</span><span class="p">[</span><span class="s1">&#39;sgrna_sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span> <span class="o">+</span>
                                               <span class="n">utr5_tiling_library</span><span class="p">[</span><span class="s1">&#39;sgrna_sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">crispick_designs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s1">&#39;data/example_cas9_sgrna-designs.txt&#39;</span><span class="p">)</span>
<span class="k">assert</span> <span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">crispick_designs</span><span class="p">[</span><span class="s1">&#39;sgRNA Sequence&#39;</span><span class="p">])</span> <span class="o">-</span> <span class="p">{</span><span class="s1">&#39;TAGGGCATTCCAGAAAGATG&#39;</span><span class="p">}</span> <span class="o">==</span>
        <span class="nb">set</span><span class="p">(</span><span class="n">cds_tiling_library</span><span class="p">[</span><span class="s1">&#39;sgrna_sequence&#39;</span><span class="p">]))</span>
<span class="n">cds_tiling_library</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>context_sequence</th>
      <th>pam_sequence</th>
      <th>sgrna_sequence</th>
      <th>sgrna_global_start</th>
      <th>sgrna_global_17</th>
      <th>sgrna_global_18</th>
      <th>sgrna_strand</th>
      <th>object_type</th>
      <th>transcript_strand</th>
      <th>transcript_id</th>
      <th>chromosome</th>
      <th>region_id</th>
      <th>region_start</th>
      <th>region_end</th>
      <th>flag</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>15</th>
      <td>GCTGTCTTTATATTCATGACCTACTGGCAT</td>
      <td>TGG</td>
      <td>TCTTTATATTCATGACCTAC</td>
      <td>5456130.0</td>
      <td>5456146.0</td>
      <td>5456147.0</td>
      <td>1</td>
      <td>Exon</td>
      <td>1</td>
      <td>ENST00000381577</td>
      <td>9</td>
      <td>ENSE00000813155</td>
      <td>5456100</td>
      <td>5456165</td>
      <td></td>
    </tr>
    <tr>
      <th>16</th>
      <td>CATGACCTACTGGCATTTGCTGAACGGTAA</td>
      <td>CGG</td>
      <td>ACCTACTGGCATTTGCTGAA</td>
      <td>5456144.0</td>
      <td>5456160.0</td>
      <td>5456161.0</td>
      <td>1</td>
      <td>Exon</td>
      <td>1</td>
      <td>ENST00000381577</td>
      <td>9</td>
      <td>ENSE00000813155</td>
      <td>5456100</td>
      <td>5456165</td>
      <td></td>
    </tr>
    <tr>
      <th>17</th>
      <td>TCTTACCGTTCAGCAAATGCCAGTAGGTCA</td>
      <td>AGG</td>
      <td>ACCGTTCAGCAAATGCCAGT</td>
      <td>5456167.0</td>
      <td>5456151.0</td>
      <td>5456150.0</td>
      <td>-1</td>
      <td>Exon</td>
      <td>1</td>
      <td>ENST00000381577</td>
      <td>9</td>
      <td>ENSE00000813155</td>
      <td>5456100</td>
      <td>5456165</td>
      <td></td>
    </tr>
    <tr>
      <th>19</th>
      <td>CTTTCTTTTTAGCATTTACTGTCACGGTTC</td>
      <td>CGG</td>
      <td>CTTTTTAGCATTTACTGTCA</td>
      <td>5457071.0</td>
      <td>5457087.0</td>
      <td>5457088.0</td>
      <td>1</td>
      <td>Exon</td>
      <td>1</td>
      <td>ENST00000381577</td>
      <td>9</td>
      <td>ENSE00003541119</td>
      <td>5457079</td>
      <td>5457420</td>
      <td>TTTT</td>
    </tr>
    <tr>
      <th>20</th>
      <td>TAGCATTTACTGTCACGGTTCCCAAGGACC</td>
      <td>AGG</td>
      <td>ATTTACTGTCACGGTTCCCA</td>
      <td>5457080.0</td>
      <td>5457096.0</td>
      <td>5457097.0</td>
      <td>1</td>
      <td>Exon</td>
      <td>1</td>
      <td>ENST00000381577</td>
      <td>9</td>
      <td>ENSE00003541119</td>
      <td>5457079</td>
      <td>5457420</td>
      <td></td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>743</th>
      <td>TCTCCAAAAACACCCATCCAGGCAGGGGGA</td>
      <td>GGG</td>
      <td>CAAAAACACCCATCCAGGCA</td>
      <td>140734646.0</td>
      <td>140734630.0</td>
      <td>140734629.0</td>
      <td>-1</td>
      <td>Exon</td>
      <td>-1</td>
      <td>ENST00000644969</td>
      <td>7</td>
      <td>ENSE00001907699</td>
      <td>140734617</td>
      <td>140734770</td>
      <td></td>
    </tr>
    <tr>
      <th>744</th>
      <td>CTCCAAAAACACCCATCCAGGCAGGGGGAT</td>
      <td>GGG</td>
      <td>AAAAACACCCATCCAGGCAG</td>
      <td>140734645.0</td>
      <td>140734629.0</td>
      <td>140734628.0</td>
      <td>-1</td>
      <td>Exon</td>
      <td>-1</td>
      <td>ENST00000644969</td>
      <td>7</td>
      <td>ENSE00001907699</td>
      <td>140734617</td>
      <td>140734770</td>
      <td></td>
    </tr>
    <tr>
      <th>745</th>
      <td>TCCAAAAACACCCATCCAGGCAGGGGGATA</td>
      <td>GGG</td>
      <td>AAAACACCCATCCAGGCAGG</td>
      <td>140734644.0</td>
      <td>140734628.0</td>
      <td>140734627.0</td>
      <td>-1</td>
      <td>Exon</td>
      <td>-1</td>
      <td>ENST00000644969</td>
      <td>7</td>
      <td>ENSE00001907699</td>
      <td>140734617</td>
      <td>140734770</td>
      <td></td>
    </tr>
    <tr>
      <th>746</th>
      <td>AACACCCATCCAGGCAGGGGGATATGGTGC</td>
      <td>TGG</td>
      <td>CCCATCCAGGCAGGGGGATA</td>
      <td>140734638.0</td>
      <td>140734622.0</td>
      <td>140734621.0</td>
      <td>-1</td>
      <td>Exon</td>
      <td>-1</td>
      <td>ENST00000644969</td>
      <td>7</td>
      <td>ENSE00001907699</td>
      <td>140734617</td>
      <td>140734770</td>
      <td></td>
    </tr>
    <tr>
      <th>1028</th>
      <td>GCTGCCATGATGGTGGCTACTTGAAGGCTG</td>
      <td>AGG</td>
      <td>CCATGATGGTGGCTACTTGA</td>
      <td>140726482.0</td>
      <td>140726498.0</td>
      <td>140726499.0</td>
      <td>1</td>
      <td>Exon</td>
      <td>-1</td>
      <td>ENST00000644969</td>
      <td>7</td>
      <td>ENSE00003819130</td>
      <td>140719337</td>
      <td>140726516</td>
      <td></td>
    </tr>
  </tbody>
</table>
<p>401 rows  15 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

