---

title: ensembl


keywords: fastai
sidebar: home_sidebar

summary: "Query sequence data in ensembl"
description: "Query sequence data in ensembl"
nb_path: "00_ensembl.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 00_ensembl.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Query-the-ensembl-database">Query the ensembl database<a class="anchor-link" href="#Query-the-ensembl-database"> </a></h2><p>Functions to query <a href="https://rest.ensembl.org/">https://rest.ensembl.org/</a></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="GET">GET<a class="anchor-link" href="#GET"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="ensembl_get" class="doc_header"><code>ensembl_get</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/ensembl.py#L11" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>ensembl_get</code>(<strong><code>ext</code></strong>, <strong><code>query</code></strong>=<em><code>None</code></em>, <strong><code>headers</code></strong>=<em><code>None</code></em>, <strong><code>params</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Generic wrapper for using GET requests to the ensembl rest API</p>
<p>ext: str, url extension |
query: str or None, end of url extension specifying species, taxon, esnembl_id etc |
headers: dict or None,  meta-information for query |
params: dict or None, parameters for query |</p>
<p>returns: Response object</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="POST">POST<a class="anchor-link" href="#POST"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="ensembl_post" class="doc_header"><code>ensembl_post</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/ensembl.py#L34" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>ensembl_post</code>(<strong><code>ext</code></strong>, <strong><code>data</code></strong>, <strong><code>headers</code></strong>=<em><code>None</code></em>, <strong><code>params</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Generic wrapper for using POST requests to the ensembl rest API</p>
<p>ext: str, url extension |
data: dict, query data |
headers: dict or None,  meta-information for query |
params: dict or None, parameters for query |</p>
<p>returns: Response object</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Get-Transcript-Meta-information">Get Transcript Meta-information<a class="anchor-link" href="#Get-Transcript-Meta-information"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_ensembl_id_information" class="doc_header"><code>get_ensembl_id_information</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/ensembl.py#L56" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_ensembl_id_information</code>(<strong><code>ensembl_id</code></strong>, <strong><code>expand</code></strong>=<em><code>1</code></em>, <strong><code>utr</code></strong>=<em><code>1</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Get information for an ensembl id (ref. <a href="https://rest.ensembl.org/documentation/info/lookup">https://rest.ensembl.org/documentation/info/lookup</a>)</p>
<p>ensembl_id: ENSG or ENST
expand: Expands the search to include any connected features.
    e.g. If the object is a gene, its transcripts, translations and exons will be returned as well.
utr: Include 5' and 3' UTR features. Only available if the expand option is used.
kwargs: other parameters for lookup |</p>
<p>returns: dict</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="n">get_ensembl_id_information</span><span class="p">(</span><span class="s1">&#39;ENSG00000157764&#39;</span><span class="p">)[</span><span class="s1">&#39;display_name&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;BRAF&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Get-sequence-by-transcript">Get sequence by transcript<a class="anchor-link" href="#Get-sequence-by-transcript"> </a></h2><p>Note that <a href="/sgrna_designer/ensembl.html#get_transcript_sequence"><code>get_transcript_sequence</code></a> returns sequences 5' to 3' on same strand as the target</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_transcript_sequence" class="doc_header"><code>get_transcript_sequence</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/ensembl.py#L75" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_transcript_sequence</code>(<strong><code>ensembl_id</code></strong>, <strong><code>seq_type</code></strong>=<em><code>'cds'</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Request multiple types of sequence by stable identifier. Supports feature masking and expand options.
Uses <a href="https://rest.ensembl.org/documentation/info/sequence_id">https://rest.ensembl.org/documentation/info/sequence_id</a></p>
<p>seq_type: str, one of [genomic, cds, cdna, protein] |
mask_feature: str, If sequence is genomic, mask introns. If sequence is cDNA, mask UTRs. |
kwargs: additional parameter arguments |</p>
<p>returns: str, sequence 3' to 5' in the same orientation as the input transcript</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span><span class="p">(</span><span class="n">get_transcript_sequence</span><span class="p">(</span><span class="s1">&#39;ENSE00001154485&#39;</span><span class="p">,</span> <span class="n">seq_type</span><span class="o">=</span><span class="s1">&#39;genomic&#39;</span><span class="p">,</span> <span class="n">mask_feature</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                               <span class="n">expand_5prime</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span> <span class="o">==</span>
       <span class="s1">&#39;CCCTCCCCGCCCGACAGCGGCCGCTCGGGCCCCGGCTCTCGGTTATAAGATGGCGGCGCTGAGCGGTGGCGGTGGTGGCGGCGCGGAGCCGGGCCAGGCTCTGTTCAACGGGGACATGGAGCCCGAGGCCGGCGCCGGCGCCGGCGCCGCGGCCTCTTCGGCTGCGGACCCTGCCATTCCGGAGGAG&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Post-sequence-by-transcript">Post sequence by transcript<a class="anchor-link" href="#Post-sequence-by-transcript"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="post_transcript_sequence" class="doc_header"><code>post_transcript_sequence</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/ensembl.py#L94" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>post_transcript_sequence</code>(<strong><code>ensembl_ids</code></strong>, <strong><code>seq_type</code></strong>=<em><code>'cdna'</code></em>, <strong><code>mask_feature</code></strong>=<em><code>'1'</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Request multiple types of sequence by stable identifier. Supports feature masking and expand options.
Uses <a href="https://rest.ensembl.org/documentation/info/sequence_id">https://rest.ensembl.org/documentation/info/sequence_id</a></p>
<p>seq_type: str, one of [genomic, cds, cdna, protein] |
mask_feature: str, If sequence is genomic, mask introns. If sequence is cDNA, mask UTRs. |
kwargs: additional parameter arguments |</p>
<p>returns: str, sequence 3' to 5' in the same orientation as the input transcript</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span><span class="p">(</span><span class="n">post_transcript_sequence</span><span class="p">([</span><span class="s2">&quot;ENSG00000157764&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSG00000248378&quot;</span><span class="p">],</span> <span class="n">seq_type</span><span class="o">=</span><span class="s1">&#39;genomic&#39;</span><span class="p">,</span>
                                <span class="n">mask_feature</span><span class="o">=</span><span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;seq&#39;</span><span class="p">][:</span><span class="mi">4</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;CTTC&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Get-Sequence-by-region">Get Sequence by region<a class="anchor-link" href="#Get-Sequence-by-region"> </a></h2><p>Note that <a href="/sgrna_designer/ensembl.html#get_region_sequence"><code>get_region_sequence</code></a> returns sequences 5' to 3' on the positive DNA strand</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="create_region_str" class="doc_header"><code>create_region_str</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/ensembl.py#L114" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>create_region_str</code>(<strong><code>start</code></strong>, <strong><code>end</code></strong>, <strong><code>seq_region</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_region_sequence" class="doc_header"><code>get_region_sequence</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/ensembl.py#L119" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_region_sequence</code>(<strong><code>start</code></strong>, <strong><code>end</code></strong>, <strong><code>seq_region</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Returns the genomic sequence of the specified region of the given species.
Supports feature masking and expand options.
Uses <a href="https://rest.ensembl.org/documentation/info/sequence_region">https://rest.ensembl.org/documentation/info/sequence_region</a></p>
<p>start: int, genomic start position |
end: int, genomic end position |
seq_region: str, chromosome |
kwargs: additional arguments to send to query |</p>
<p>returns: str, sequence 3' to 5' on the positive DNA strand</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="p">(</span><span class="n">get_region_sequence</span><span class="p">(</span><span class="mi">1000000</span><span class="p">,</span> <span class="mi">1000100</span><span class="p">,</span> <span class="s1">&#39;X&#39;</span><span class="p">)</span> <span class="o">==</span>
        <span class="s1">&#39;CTGTAGAAACATTAGCCTGGCTAACAAGGTGAAACCCCATCTCTACTAACAATACAAAATATTGGTTGGGCGTGGTGGCGGGTGCTTGTAATCCCAGCTAC&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Post-sequence-by-region">Post sequence by region<a class="anchor-link" href="#Post-sequence-by-region"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="post_region_sequences" class="doc_header"><code>post_region_sequences</code><a href="https://github.com/gpp-rnd/sgrna_designer/tree/master/sgrna_designer/ensembl.py#L140" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>post_region_sequences</code>(<strong><code>start</code></strong>, <strong><code>end</code></strong>, <strong><code>seq_region</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Request multiple types of sequence by a list of regions.
Uses <a href="https://rest.ensembl.org/documentation/info/sequence_region_post">https://rest.ensembl.org/documentation/info/sequence_region_post</a></p>
<p>start: list, start of each query region. Should align with end and seq_region |
end: list |
seq_region: list |
kwargs: optional params for the query |</p>
<p>returns: list of dict</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="p">(</span><span class="n">post_region_sequences</span><span class="p">([</span><span class="mi">1000000</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">1000100</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span>
                              <span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;ABBA01004489.1&#39;</span><span class="p">])[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;seq&#39;</span><span class="p">]</span> <span class="o">==</span>
        <span class="s1">&#39;CTGTAGAAACATTAGCCTGGCTAACAAGGTGAAACCCCATCTCTACTAACAATACAAAATATTGGTTGGGCGTGGTGGCGGGTGCTTGTAATCCCAGCTAC&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

